lib/internal/modules/esm/module_map.js

```diff
@@ -1,17 +1,119 @@
 'use strict';
 
-const { kImplicitAssertType } = require('internal/modules/esm/assert');
 const {
+  ArrayPrototypeJoin,
+  ArrayPrototypeMap,
+  ArrayPrototypeSort,
+  JSONStringify,
+  ObjectDefineProperty,
+  ObjectKeys,
   SafeMap,
 } = primordials;
+const { kImplicitAssertType } = require('internal/modules/esm/assert');
+const { setOwnProperty } = require('internal/util');
 let debug = require('internal/util/debuglog').debuglog('esm', (fn) => {
   debug = fn;
 });
 const { ERR_INVALID_ARG_TYPE } = require('internal/errors').codes;
 const { validateString } = require('internal/validators');
 
-// Tracks the state of the loader-level module cache
-class ModuleMap extends SafeMap {
+/**
+ * Cache the results of the `resolve` step of the module resolution and loading process.
+ * Future resolutions of the same input (specifier, parent URL and import assertions)
+ * must return the same result if the first attempt was successful, per
+ * https://tc39.es/ecma262/#sec-HostLoadImportedModule.
+ */
+class ResolveCache extends SafeMap {
+  constructor(i) { super(i); } // eslint-disable-line no-useless-constructor
+
+  /**
+   * Generates the internal serialized cache key and returns it along the actual cache object.
+   *
+   * It is exposed to allow more efficient read and overwrite a cache entry.
+   * @param {string} specifier
+   * @param {string} parentURL
+   * @param {Record<string,string>} importAssertions
+   * @returns {{
+   *   serializedModuleRequest: string,
+   *   internalCache: Record<string, Promise<import('./loader').ModuleExports> | import('./loader').ModuleExports>,
+   * }}
+   */
+  getSerialized(specifier, parentURL, importAssertions) {
+    let internalCache = super.get(parentURL);
+    if (internalCache == null) {
+      super.set(parentURL, internalCache = { __proto__: null });
+    }
+
+    // To serialize the ModuleRequest (specifier + list of import attributes),
+    // we need to sort the attributes by key, then stringifying,
+    // so that different import statements with the same attributes are always treated
+    // as identical.
```

Comment:
Are we doing anything like this serialization today? If not, wouldn’t adding it be a big performance hit? Can you run any benchmarks to see how much of an impact this has?

I’m still ambivalent as to whether this fix needs to happen at all, seeing as we’ve had no user complaints about it, but if it also incurs a noticeable performance hit then I’d be more inclined to go back to TC39 and ask that the spec be updated to make the behavior host-defined (so if browsers want to return the cached result, fine, but Node can choose otherwise).

Reply:
You're welcome to run the benchmark CI and/or complain to TC39 (but consider https://github.com/nodejs/node/pull/46662#issuecomment-1574452202 first), but I'm not sure I understand what's the point, we want to be spec-compliant before being performant.

Reply:
We're out of compliance today and there isn't a rush to get into compliance. If we contact them and a change is possible, we could just stay as we are and wait for them to update.